### ISSUE TITLE
False positive error[E0381]: used binding is possibly-uninitialized when using if let returning never

### ISSUE THREAD

**Issue by Skaytacium** [rust-lang/rust#111532]
### Code

```Rust
fn main() {
	let bytes: &'static str;
	let file: Result<&str, &str> = Ok("File data");

	if let Err(_error) = file {
		std::process::exit(1);
	} else if let Ok(store) = file {
		bytes = store
	}
	
	println!("{}", bytes);
}
```

### Current output

```Shell
error[E0381]: used binding `bytes` is possibly-uninitialized
  --> src/main.rs:11:17
   |
2  |     let bytes: &'static str;
   |         ----- binding declared here but left uninitialized
...
8  |         bytes = store
   |         ------------- binding initialized here in some conditions
...
11 |     println!("{}", bytes);
   |                    ^^^^^ `bytes` used here but it is possibly-uninitialized
   |
   = note: this error originates in the macro `$crate::format_args_nl` which comes from the expansion of the macro `println` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0381`.
error: could not compile `playground` due to previous error
```

### Desired output

```Shell
File data
```

### Other cases

```Rust
fn main() {
	let bytes: &'static str;
	let file: Result<&str, &str> = Ok("File data");

	match file {
	    Ok(store) => bytes = store,
	    Err(_err) => std::process::exit(1),
	}
	
	println!("{}", bytes);
}
```

### Anything else?

Tested on rust playground using stable rustc 1.69.0. `match`ing works.

**Comment by wwylele** [comment]
consecutive  `if let` cannot check exhaustivity like match, so your first code should be read like

```rust
fn main() {
	let bytes: &'static str;
	let file: Result<&str, &str> = Ok("File data");

	if let Err(_error) = file {
		std::process::exit(1);
	} else if let Ok(store) = file {
		bytes = store
	} else {
		// bytes uninitialized for this case
		// the compiler doesn't recognize this else block is impossible
	}
	
	println!("{}", bytes);
}

```

**Comment by Skaytacium** [comment]
Interesting, is this a known bug? I couldn't find it in this repo's issues. Apologies if this is a duplicate issue.

**Comment by wwylele** [comment]
I guess the reason it's hard to find the discussion in the issue list is because this is "known, not a bug". My intuitive says it is much harder to implement the check for if-else chain, and even harder to have a clear specification, given you can put arbitrary, possibly related or unrelated conditions in each `if` which can be impossible to reason about at compile time.

**Comment by Skaytacium** [comment]
Makes sense. I thought that when you're checking for enum variants in an `if let` chain, the exhaustivity would be considered like it is in match. I guess its better used as syntactical sugar for extracting data out of a single enum variant. Anyway, I'll close this issue. Thanks.
